lib/experimental/class-wp-rest-navigation-controller.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Navigation_Controller class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+require __DIR__ . '/class-wp-navigation-fallbacks-gutenberg.php';
+
+
+/**
+ * Base Templates REST API Controller.
+ */
+class WP_REST_Navigation_Controller extends WP_REST_Posts_Controller {
+
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+
+		parent::register_routes();
+
+		// Lists a single nav item based on the given id or slug.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/fallback',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_fallbacks' ),
+					'permission_callback' => array( $this, 'get_fallbacks_permissions_check' ),
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::READABLE ),
+				),
+				'allow_batch' => $this->allow_batch,
+				'schema'      => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to read fallbacks.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_fallbacks_permissions_check( $request ) {
+
+		$post_type = get_post_type_object( $this->post_type );
+
+		// Getting fallbacks also requires creating `wp_navigation` posts.
+		if ( ! current_user_can( $post_type->cap->create_posts ) ) {
+			return new WP_Error(
+				'rest_cannot_create',
+				__( 'Sorry, you are not allowed to create Navigation Menus as this user.', 'default' ),
+				array( 'status' => rest_authorization_required_code() )
+			);
+		}
```

Comment:
I have concerns about this. Whilst the fallback logic may create posts, this logic may need to be executed by the front of the site (as a one time operation).

Therefore if we only allow this for logged in users then fallbacks will not work. Moreover, the permission levels required for create seem to be pretty high meaning only `administrator` roles will be able to do this.

Why do posts sometimes need to be created by this endpoint? Because  a user may activate a theme which contains an "empty" Navigation block (one without a `ref` attribute pointing to a `wp_navigation` post which contains the menu items). In this scenario we need to retrieve a fallback and the fallback logic dictates that if a Nav Menu does not exist we should attempt to create on for the user. Therefore if the front of the site is visited prior to the Nav block being configured then we need to run the fallback mechanic on the front of the site which (as a one time operation) _may_ create a `wp_navigation` post.

Reply:
Can we change the capabilities to allow anyone to call it? If not then I guess the block can call the function directly in the front end. Don't forget that the block can be added to the post editor too, by non-admins.

Reply:
The block can be added but I don't think you can save the navigation menus. There are restrictions which is why we have the `canUser` calls in the `edit.js` editor code.

Perhaps we need to reconsider actually creating a menu when users hit the front of the site. Perhaps rather the front end code should simply return the fallback _blocks_ directly (e.g. page list). 

I think this might be a good approach because as soon as the user opens the editor the fallback process would run and create the most appropriate menu. Up until that point the front of the site would just show a Page List which is a suitable experience.

This would avoid 

- concerns about write on render
- concerns about permissions with write for non-logged in/users with lower permissions.

What do you think?

Reply:
@scruffian @draganescu I'd be keen to hear your thoughts here 🙏

Reply:
Why does creating a fallback menu by lower permissions users constitute a problem? As a fallback it more or less like placeholder content. It doesn't matter who and when creates the fallback menu, be it a page list block menu or a converted classic menu. The goal is to avoid an empty state.

Reply:
> What do you think?

One of the aims of this PR is to consolidate the logic between back and front end, but this suggestion seems to move away from that goal.

Reply:
That is accurate in a general sense. Specifically however, this PR still consolidates several pieces of code that were split between JS and PHP.

The specifics of how the menu is retrieved for the front end is the main change I'm thinking about. This would still mean editor and front end share (a lot of) code, but that no "writes" are ever done on the front end.

[My suggestion](https://github.com/WordPress/gutenberg/pull/48698#discussion_r1147660921) is still up for debate. I'm trying to come at this from the perspective of:

- what is the _correct_ implementation permissions/security wise?
- what would we do if we were starting from scratch?

I think next steps are to get some input from Core folks on permissions angle. With that we can make more informed decisions.

Reply:
I'd just realised that the permissions on this endpoint are nothing at all to do with what happens on the front of the site 🤦  

So we just need to be happy that the permissions are suitable for editor users of varying permissions.